cache/refs.go

```diff
@@ -90,6 +97,10 @@ type cacheRecord struct {
 	equalImmutable *immutableRef
 
 	parentChainCache []digest.Digest
+
+	// TODO clear out entry in map when session ends?
+	// map of sessionID -> provider and/or progress writer that should be used
+	perSessionState map[string]perSessionState
```

Comment:
It would grow until the cacheRecord is removed from memory, which I believe only happens if it's pruned or if buildkitd shuts down (the map isn't persisted to disk, so it won't get reloaded into memory when buildkitd starts again).

I left the TODO because of this concern about the map growing potentially indefinitely if many sessions reference the same cache ref. I think it can be fixed by using the [`Context`](https://github.com/moby/buildkit/blob/6f83882a0b098034817f37858c32d45133c5703c/session/manager.go#L194) of the session object to clear the map when that context gets canceled. This would require passing the session manager one way or another to the cache manager (right now session manager is only provided per vertex, not to the worker, but maybe that can change or the session manager could just be provided as another `perSessionState` option). Let me know if there's any other suggestions though.